TraceRecorder demo added in CORTEX_MPS2_QEMU_IAR_GCC

johankraft opened the pull request:
TraceRecorder demo added in CORTEX_MPS2_QEMU_IAR_GCC

Description
-----------
The CORTEX_MPS2_QEMU_IAR_GCC demo has been extended with Percepio TraceRecorder. The steps needed to replicate the setup in a different project is documented as "TODO" code comments, that shows up in the Eclipse "Tasks" list.

Test Steps
-----------
Run the project in QEMU, like before. Save the trace using the debugger (see below) and open the resulting file in Tracealyzer or View. To save the trace using gdb (e.g. from Eclipse), run the gdb command "dump binary value trace.bin *RecorderDataPtr" in the debugger console. To save TraceRecorder snapshots using IAR, see https://percepio.com/iar.
See also "Usage" sections added in FreeRTOS/Demo/CORTEX_MPS2_QEMU_IAR_GCC/Readme.md

Checklist:
----------
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x ] I have tested my changes. No regression in existing tests.
- [ ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------
<!-- If any, please provide issue ID. -->


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

johankraft commented:
I have only tested manually, not using any CI tests in FreeRTOS. But there are no changed in the kernel, only in the demo code.

felicityzhao9 commented:
Thanks for your demo contribution Johan! Our team will take a look and get in touch with you soon!

amazonKamath commented:
@johankraft  Is the TraceRecorderSource  here same as https://github.com/percepio/TraceRecorderSource/tree/main?

johankraft commented:
Yes, it is the same TraceRecorder code. I think you already have it in the
tree somewhere.
But I included a new copy of TraceRecorder as a subdirectory instead, to
make it easier to find in the demo. We can change this if you prefer.

/Johan

Den lör 18 jan. 2025 06:42Nikhil Kamath ***@***.***> skrev:

> @johankraft <https://github.com/johankraft> Is the TraceRecorderSource
> here same as https://github.com/percepio/TraceRecorderSource/tree/main?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeRTOS/FreeRTOS/pull/1320#issuecomment-2599564793>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APLPNIM76RJJU52LW3FU5632LHSVTAVCNFSM6AAAAABVMAM2DOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJZGU3DINZZGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

amazonKamath commented:
> Yes, it is the same TraceRecorder code. I think you already have it in the tree somewhere. But I included a new copy of TraceRecorder as a subdirectory instead, to make it easier to find in the demo. We can change this if you prefer. /Johan Den lör 18 jan. 2025 06:42Nikhil Kamath ***@***.***> skrev:
> […](#)
> @johankraft <https://github.com/johankraft> Is the TraceRecorderSource here same as https://github.com/percepio/TraceRecorderSource/tree/main? — Reply to this email directly, view it on GitHub <[#1320 (comment)](https://github.com/FreeRTOS/FreeRTOS/pull/1320#issuecomment-2599564793)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/APLPNIM76RJJU52LW3FU5632LHSVTAVCNFSM6AAAAABVMAM2DOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJZGU3DINZZGM> . You are receiving this because you were mentioned.Message ID: ***@***.***>

I would let others weigh in here as well, I think we should go for git submodule to the TraceRecorder upstream instead. Thoughts?

johankraft commented:
Using a submodule can of course be an alternative. We can do that if you
prefer.

Note however that the TraceRecorder repo on Github contains "kernelport"
files for other RTOS as well (see the kernelports folder). This was partly
why I added the necessary files locally. Not sure if you see that as a
problem.

/Johan

Den lör 18 jan. 2025 12:20Nikhil Kamath ***@***.***> skrev:

> Yes, it is the same TraceRecorder code. I think you already have it in the
> tree somewhere. But I included a new copy of TraceRecorder as a
> subdirectory instead, to make it easier to find in the demo. We can change
> this if you prefer. /Johan Den lör 18 jan. 2025 06:42Nikhil Kamath *@*.
>
> *> skrev: … <#m_-6264774252083051326_> @johankraft
> <https://github.com/johankraft> https://github.com/johankraft
> <https://github.com/johankraft> Is the TraceRecorderSource here same as
> https://github.com/percepio/TraceRecorderSource/tree/main
> <https://github.com/percepio/TraceRecorderSource/tree/main>? — Reply to
> this email directly, view it on GitHub <#1320 (comment)
> <https://github.com/FreeRTOS/FreeRTOS/pull/1320#issuecomment-2599564793>>,
> or unsubscribe
> https://github.com/notifications/unsubscribe-auth/APLPNIM76RJJU52LW3FU5632LHSVTAVCNFSM6AAAAABVMAM2DOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJZGU3DINZZGM
> <https://github.com/notifications/unsubscribe-auth/APLPNIM76RJJU52LW3FU5632LHSVTAVCNFSM6AAAAABVMAM2DOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJZGU3DINZZGM>
> . You are receiving this because you were mentioned.Message ID: @.*>
>
> I would let others weigh in here as well, I think we should go for git
> submodule to the TraceRecorder upstream instead. Thoughts?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeRTOS/FreeRTOS/pull/1320#issuecomment-2599676952>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APLPNIKI67WQ77ESG65L4ST2LI2HPAVCNFSM6AAAAABVMAM2DOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJZGY3TMOJVGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

amazonKamath commented:
@aggarg  thoughts?

aggarg commented:
@johankraft Thank you for your contribution. TraceRecorder is already sub-moduled at `FreeRTOS-Plus/Source/FreeRTOS-Plus-Trace`. Please use the same in this project. You may need to update the submodule pointer.

Keeping a separate copy is hard to maintain as it goes outdated really fast.

johankraft commented:
Sure, we can use that instead. But FreeRTOS-Plus-Trace seems to point at an
old version of TraceRecorder, from two years ago. That is unlikely to work
with the latest FreeRTOS kernel. Is that correct? Doesn't the submodule
link to the latest version? If so, it needs to be updated.
Or perhaps I made some mistake when cloning the repo?

/Johan


On Mon, Jan 20, 2025 at 12:25 PM Gaurav-Aggarwal-AWS <
***@***.***> wrote:

> @johankraft <https://github.com/johankraft> Thank you for your
> contribution. TraceRecorder is already sub-moduled at
> FreeRTOS-Plus/Source/FreeRTOS-Plus-Trace. Please use the same in this
> project. You may need to update the submodule pointer.
>
> Keeping a separate copy is hard to maintain as it goes outdated really
> fast.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeRTOS/FreeRTOS/pull/1320#issuecomment-2602163036>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APLPNINMXZI7V3V224RZCR32LTMJBAVCNFSM6AAAAABVMAM2DOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMBSGE3DGMBTGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Best Regards,

Johan Kraft
Percepio AB

www.percepio.com

johankraft commented:
Update - Sorry, I missed the part about updating the submodule pointer. I have done that now and updated the makefile to use the FreeRTOS-Plus-Trace submodule instead of static files. Added the TraceRecorderConfig folder to keep local versions of the project-specific configuration files. Please have a look now (four commits in total).

aggarg commented:
I think you still need to delete the local copy of TraceRecorder.

johankraft commented:
Ops. I deleted it i. my local repo, but must have failed to commit that
part.

Den mån 20 jan. 2025 18:03Gaurav-Aggarwal-AWS ***@***.***>
skrev:

> I think you still need to delete the local copy of TraceRecorder.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeRTOS/FreeRTOS/pull/1320#issuecomment-2602909146>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APLPNIJX7FAKW2SI55LQTPL2LUT4RAVCNFSM6AAAAABVMAM2DOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMBSHEYDSMJUGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

johankraft commented:
Should be sorted now. The TraceRecorder folder is gone.

aggarg commented on FreeRTOS/Demo/CORTEX_MPS2_QEMU_IAR_GCC/build/gcc/.project:
```diff
@@ -35,6 +35,11 @@
+		<link>
+			<name>TraceRecorder</name>
+			<type>2</type>
+			<locationURI>PARENT-2-PROJECT_LOC/TraceRecorder</locationURI>
```
Does this need to be updated to `FreeRTOS-Plus-Trace`?

kar-rahul-aws commented on FreeRTOS/Demo/CORTEX_MPS2_QEMU_IAR_GCC/main_blinky.c:
```diff
@@ -113,6 +113,9 @@ void main_blinky( void )
     xQueue = xQueueCreate( mainQUEUE_LENGTH, sizeof( uint32_t ) );
 
+    /* TODO TraceRecorder (Tweak 4): Setting a name for the queue (optional). */
+    vTraceSetQueueName(xQueue, "Blinky-Queue");
```
Same comment as above - add the preprocessor directive guards to avoid compilation warnings and errors.

kar-rahul-aws commented on FreeRTOS/Demo/CORTEX_MPS2_QEMU_IAR_GCC/Readme.md:
````diff
@@ -11,3 +11,30 @@
+```
+Note that you can copy/paste this command into the Eclipse Debugger Console by using Ctrl-C, Ctrl-V.
+
+### Usage with IAR Embedded Workbench for Arm
````
The IAR demo with the trace enabled, does not seem to compile for now. It would be better to mention here it is a work in progress.

johankraft replied:
Good point. My IAR license has expired, but should receive a new one soon. I should be able to add IAR support soon, but I can update the readme with "work in progress" for now.

johankraft replied:
Updated readme about IAR support. Added a brief explanation how a user can extend the IAR project with TraceRecorder. See latest commit. I expect to update the IAR project soon, within a few days hopefully, but you never know what tomorrow brings...

kar-rahul-aws replied:
lgtm.

johankraft commented:
Regarding adding #if (configUSE_TRACE_FACILITY == 1), note that TraceRecorder has been designed to avoid this need. As long as #include "trcRecorder.h" remains, setting configUSE_TRACE_FACILITY to 0 excludes all TraceRecorder code and TraceRecorder API calls are replaced by empty #defines. However, in this case, the function xTraceTimestampSetPeriod() was missing such a define. This caused the build error. I had already fixed this issue in my local copy of TraceRecorder, that was removed, but it is not yet fixed in the main TraceRecorder repo submoduled as FreeRTOS-Plus-Trace.

I propose that we only add an #if guard block only around the three TraceRecorder calls in main() and add a comment to explain it. It is not needed in main_blinky.c.

I have fixed the invalid reference to the old TraceRecorder folder the Eclipse project. Simply removed it. Added a link to the TraceRecorderConfig folder that is more important to have easily available.

New commit pushed.

kar-rahul-aws commented on FreeRTOS/Demo/CORTEX_MPS2_QEMU_IAR_GCC/FreeRTOSConfig.h:
```diff
@@ -133,4 +135,7 @@
 #define configENABLE_BACKWARD_COMPATIBILITY 0
 
+/* TODO TraceRecorder (Step 5): Include trcRecorder.h at the end of FreeRTOSConfig.h. */
+#include "trcRecorder.h"
```
Resolved in the thread below.

johankraft commented:
I don't quite understand the failed CI checks. Anything I should address here? At least some things seem unrelated (e.g. MSP430). What is the "Check File Headers" error about? Should I exclude the TraceRecorder header files in .github/scripts/core_checker.py?

aggarg commented:
@johankraft I fixed those checks for you.

johankraft commented:
Great, thanks! So all completed now? No need to update the manifest? (it looks OK, version: "Tz4/4.10/v4.10.2")

aggarg commented:
Yes, all completed. I did that change too and merged the PR. Thank you for your contribution!

johankraft commented:
Thanks for your assistance!

Den tors 23 jan. 2025 08:31Gaurav-Aggarwal-AWS ***@***.***>
skrev:

> Yes, all completed. I did that change too and merged the PR. Thank you for
> your contribution!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeRTOS/FreeRTOS/pull/1320#issuecomment-2609051732>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APLPNIPU6ARIFOYQCVSMZQD2MCLGXAVCNFSM6AAAAABVMAM2DOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMBZGA2TCNZTGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
